File: src/main/java/org/minutebots/robot/hardwareconfigurations/SimulatedBot.java
package org.minutebots.robot.hardwareconfigurations;

import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.wpilibj.SpeedController;
import edu.wpi.first.wpilibj.shuffleboard.BuiltInWidgets;
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
import org.minutebots.lib.VirtualMotor;

public class SimulatedBot implements HardwareManger {
    private NetworkTableEntry piston = Shuffleboard.getTab("Virtual Motors")
            .add("Intake Piston", false).getEntry(),
            gyro = Shuffleboard.getTab("Virtual Motors")
                    .add("Gyro Angle", 0)
                    .withWidget(BuiltInWidgets.kTextView)
                    .getEntry(),
            topLimit = Shuffleboard.getTab("Virtual Motors")
                    .add("Top Limit Switch", false)
                    .withWidget(BuiltInWidgets.kToggleButton)
                    .getEntry(),
            bottomLimit = Shuffleboard.getTab("Virtual Motors")
                    .add("Bottom Limit Switch", false)
                    .withWidget(BuiltInWidgets.kToggleButton)
                    .getEntry();

    private SpeedController[] motors = new SpeedController[]{
            new VirtualMotor(Asimov.LEFT_FRONT_MOTOR),
            new VirtualMotor(Asimov.LEFT_BACK_MOTOR),
            new VirtualMotor(Asimov.RIGHT_FRONT_MOTOR),
            new VirtualMotor(Asimov.RIGHT_BACK_MOTOR)};

    private VirtualMotor arm = new VirtualMotor(Asimov.DEPOT_ARM),
            roller = new VirtualMotor(Asimov.DEPOT_WHEEL),
            ramp = new VirtualMotor(Asimov.RAMP);

    @Override
    public SpeedController[] drivetrainMotors() {
        return motors;
    }

    @Override
    public void resetGyro() {
        gyro.setDouble(0);
    }

    @Override
    public SpeedController depotArm() {
        return arm;
    }

    @Override
    public SpeedController depotRoller() {
        return roller;
    }

    @Override
    public SpeedController rampMotor() {
        return ramp;
    }

    @Override
    public void extendIntakePiston() {
        piston.setBoolean(true);
    }

    @Override
    public void retractIntakePiston() {
        piston.setBoolean(false);
    }

    @Override
    public void closeSolenoids() {
        piston.setBoolean(false);
    }

    @Override
    public double getAngle() {
        return -gyro.getDouble(0);
    }

    @Override
    public double ultraSonicDist() {
        return 0;
    }

    @Override
    public boolean armUp() {
        return topLimit.getBoolean(false);
    }

    @Override
    public boolean armDown() {
        return bottomLimit.getBoolean(false);
    }

    @Override
    public void init() {

    }
}

File: src/main/java/org/minutebots/robot/hardwareconfigurations/SimulatedBot.kt
package org.minutebots.robot.hardwareconfigurations

import edu.wpi.first.networktables.NetworkTableEntry
import edu.wpi.first.wpilibj.SpeedController
import edu.wpi.first.wpilibj.shuffleboard.BuiltInWidgets
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard
import org.minutebots.lib.VirtualMotor

class SimulatedBot : HardwareManger {
    private val piston = Shuffleboard.getTab("Virtual Motors")
            .add("Intake Piston", false).entry
    private val hatchm = Shuffleboard.getTab("Virtual Motors")
            .add("Active Hatch", false).entry
    private val ramplock = Shuffleboard.getTab("Virtual Motors")
            .add("Ramp Lock", false).entry
    private val gyro = Shuffleboard.getTab("Virtual Motors")
            .add("Gyro Angle", 0)
            .withWidget(BuiltInWidgets.kTextView)
            .entry
    private val topLimit = Shuffleboard.getTab("Virtual Motors")
            .add("Top Limit Switch", false)
            .withWidget(BuiltInWidgets.kToggleButton)
            .entry
    private val bottomLimit = Shuffleboard.getTab("Virtual Motors")
            .add("Bottom Limit Switch", false)
            .withWidget(BuiltInWidgets.kToggleButton)
            .entry

    private val motors = arrayOf<SpeedController>(VirtualMotor(Asimov.LEFT_FRONT_MOTOR), VirtualMotor(Asimov.LEFT_BACK_MOTOR), VirtualMotor(Asimov.RIGHT_FRONT_MOTOR), VirtualMotor(Asimov.RIGHT_BACK_MOTOR))

    private val arm = VirtualMotor(Asimov.DEPOT_ARM)
    private val roller = VirtualMotor(Asimov.DEPOT_WHEEL)
    private val ramp = VirtualMotor(Asimov.RAMP)

    override val angle: Double
        get() = -gyro.getDouble(0.0)

    override fun drivetrainMotors(): Array<SpeedController> {
        return motors
    }

    override fun resetGyro() {
        gyro.setDouble(0.0)
    }

    override fun depotArm(): SpeedController {
        return arm
    }

    override fun depotRoller(): SpeedController {
        return roller
    }

    override fun rampMotor(): SpeedController {
        return ramp
    }

    override fun extendIntakePiston() {
        piston.setBoolean(true)
    }

    override fun retractIntakePiston() {
        piston.setBoolean(false)
    }

    override fun closeSolenoids() {
        piston.setBoolean(false)
    }

    override fun ultraSonicDist(): Double {
        return 0.0
    }

    override fun armUp(): Boolean {
        return topLimit.getBoolean(false)
    }

    override fun armDown(): Boolean {
        return bottomLimit.getBoolean(false)
    }

    override fun extendActiveHatch() {
        hatchm.setBoolean(true)
    }

    override fun retractActiveHatch() {
        hatchm.setBoolean(false)
    }

    override fun extendRampLock() {
        ramplock.setBoolean(true)
    }

    override fun retractRampLock() {
        ramplock.setBoolean(false)
    }

    override fun init() {

    }
}
